presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java

```diff
@@ -795,15 +799,86 @@ public PrestoS3InputStream(AmazonS3 s3, String host, Path path, int maxAttempts,
         @Override
         public void close()
         {
-            closed = true;
+            closed.set(true);
             closeStream();
         }
 
+        @Override
+        public int read(long position, byte[] buffer, int offset, int length)
+                throws IOException
+        {
+            checkClosed();
+            if (position < 0) {
+                throw new EOFException(FSExceptionMessages.NEGATIVE_SEEK);
+            }
+            checkPositionIndexes(offset, offset + length, buffer.length);
+            if (length == 0) {
+                return 0;
+            }
+
+            try {
+                return retry()
+                        .maxAttempts(maxAttempts)
+                        .exponentialBackoff(BACKOFF_MIN_SLEEP, maxBackoffTime, maxRetryTime, 2.0)
+                        .stopOn(InterruptedException.class, UnrecoverableS3OperationException.class, EOFException.class)
+                        .onRetry(STATS::newGetObjectRetry)
+                        .run("getS3Object", () -> {
+                            InputStream in;
+                            try {
+                                GetObjectRequest request = new GetObjectRequest(host, keyFromPath(path))
+                                        .withRange(position, (position + length) - 1);
+                                in = s3.getObject(request).getObjectContent();
```

Comment:
`in` is a object field and here we override it and finally close it. Can this and `read(byte[] buffer, int offset, int length)` be interleaved. This seems to break the other read

Reply:
It's a local variable declared at the start of this lambda. I can rename the local variable if you'd like.